internal/server/add_payment.go

```diff
@@ -51,6 +51,16 @@ func AddPayment(client AddPaymentClient, tmpl template.Template) Handler {
 						"amount": {"reason": "Please enter the amount to 2 decimal places"},
 					},
 				}
+				if data.Source == "" {
+					data.Error.Field["source"] = map[string]string{
+						"reason": "Value is required and can't be empty",
+					}
+				}
+				if data.PaymentDate == "" {
+					data.Error.Field["paymentDate"] = map[string]string{
+						"reason": "Value is required and can't be empty",
+					}
+				}
```

Comment:
I'm not a big fan of this because it duplicates validation and doesn't cover every option (e.g. if you enter a bad amount and an invalid source, it will only tell you about the amount) but I can't think of a better option right now.

Reply:
i'm with you, I don't like this at all but it was all I could think of :(

Reply:
Eurgh, the wonders of Laminas doing its filter annotations before its validation ones 🙄 Oh well, guess that means it's less of a problem 😂